docs/troubleshooting.md

```diff
@@ -0,0 +1,200 @@
+
+There are also appendices with information which may be useful for manual
+debugging.
+
+Your first steps are to ensure that you have logging so you can see what is going
+on; a level of `INFO` or even `DEBUG` will be useful.
+
+
+## Narrowing in on the problem
+
+### Check the pusher is registered in your homeserver
+
+Typically, applications will register a pusher on startup.
+
+If you have access to your homeserver, you can check that it is making it there.
+
+Start your application and then run a query against the database.
+
+#### On Synapse
+
+Use `sqlite3 /path/to/homeserver.db` or `psql synapse` as required for your
+deployment.
+
+```sql
+SELECT app_id, data FROM pushers
+WHERE user_name = '@my.user:example.org' AND kind='http';
+```
+
+You should see something like:
+
+```
+      app_id       |                        data
+-------------------+--------------------------------------------------------
+ org.example.chat  | {"format":"event_id_only",
+                   |  "url":"https://example.org/_matrix/push/v1/notify"}
+```
+
+
+#### On other homeserver implementations
+
+No details available, but contributions welcomed.
+
+
+### Check the push gateway (Sygnal) is reachable from the homeserver
+
+Following on from the example above, the homeserver's database contains the
+push gateway URL of `https://example.org/_matrix/push/v1/notify`.
+
+It may be worth manually checking that the push gateway is reachable; e.g. with
+curl:
+
+```
+$ curl https://example.org/_matrix/push/v1/notify
+<html>
+  <head><title>405 - Method Not Allowed</title></head>
+  <body>
+    <h1>Method Not Allowed</h1>
+    <p>Your browser approached me (at /_matrix/push/v1/notify) with the method "GET".  I only allow the methods HEAD, POST here.</p>
+  </body>
+</html>
+```
+
+If you get a response, such as an error like 405 Method Not Allowed, as above,
+this would suggest that the push gateway is at least reachable.
+
+If you get a **404 No Such Resource** error on the `/_matrix/push/v1/notify` endpoint,
+then chances are that your reverse proxy is not configured to pass through the
+full URL.
+
+If you don't get a HTTP response, then it is probably worth investigation.
+Check that:
+
+* Sygnal is running
+* Sygnal's configuration makes it listen on the desired port
+* Any reverse proxies are correctly set up and running
+* The firewall permits inbound traffic on the port in question.
+
+
+## Troubleshooting Firebase Notifications
+
+### iOS-specific troubles with apps using Firebase
+
+#### App doesn't receive notifications when inactive
+
+Sygnal currently only sends 'data messages' (also called 'silent notifications',
+but this name could be misleading).
+
+Whereas data messages will wake up apps on Android with no additional changes,
+iOS needs to be told that a notification is meant to wake up an inactive app.
+This is done with the `content_available` flag, which you can set in your
+`fcm_options` dictionary for the Firebase pushkin.
```

Comment:
Setting an option for APNs is done using the `fcm_options` dict?

Reply:
Yes, as confusing as that is, since you need to tell FCM to set the APNs flag.

If you have suggestions to clarify, I'm all ears.

(Tempted to reorder so APNs is first, that way it's harder for your eyes to latch on to the mention of 'iOS' in the Firebase section)

Reply:
So this section is about sending notifications to iOS through FCM?

Reply:
Yes, exactly (what a monstrosity, but people want to do it)

Reply:
any idea on setting apn_options?